File: shopapp-backend/shopapp/src/main/java/com/binhnc/shopapp/dto/OrderDTO.java
package com.binhnc.shopapp.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Size;
import lombok.*;

@Data // toString
@Builder
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class OrderDTO {
    @Min(value = 1, message = "User's ID must be > 0")
    @JsonProperty("user_id")
    private Long userId;

    @JsonProperty("fullname")
    private String fullName; // Không yêu cầu validate

    @JsonProperty("email")
    private String email; // Không yêu cầu validate

    @NotBlank(message = "Phone number is required")
    @Size(min = 5, message = "Phone number must be at last 5 characters")
    @JsonProperty("phone_number")
    private String phoneNumber; // Quan trọng

    private String address; // Không yêu cầu validate

    private String note;

    @Min(value = 0, message = "Total money must be >= 0")
    @JsonProperty("total_money")
    private Float totalMoney;

    @JsonProperty("shipping_method")
    private String shippingMethod;

    @JsonProperty("shipping_address")
    private String shippingAddress;

    @JsonProperty("payment_method")
    private String paymentMethod;

    // status, mặc định là PENDING
}

File: shopapp-angular/src/app/dtos/order/order.dto.ts
import {ArrayMinSize, IsNotEmpty, IsNumber, IsPhoneNumber, IsString} from "class-validator";

export class OrderDTO {
  @IsNumber()
  user_id: number;

  @IsString()
  fullname: string;

  @IsString()
  email: string;

  @IsPhoneNumber()
  phone_number: string;

  @IsString()
  @IsNotEmpty()
  address: string;

  @IsString()
  note: string;

  @IsNumber()
  total_money: number;
  @IsString()
  payment_method: string;

  @IsString()
  shipping_method: string;

  @IsString()
  coupon_code: string;

  // Lưu thông tin giỏ hàng
  @ArrayMinSize(1, {message: 'At least one item in cart is required'})
  cart_items: { product_id: number, quantity: number } [];

  constructor(data: any) {
    this.user_id = data.user_id;
    this.fullname = data.fullname;
    this.email = data.email;
    this.phone_number = data.phone_number;
    this.address = data.address;
    this.note = data.note;
    this.total_money = data.total_money;
    this.payment_method = data.payment_method;
    this.shipping_method = data.shipping_method;
    this.coupon_code = data.coupon_code;
    this.cart_items = data.cart_items;
  }
}
